Caching section (Cache component + caching by tags) (#54)
deity-io/falcon-docs: docs/falcon-server/caching.md

title: Caching
---

Falcon-Server provides several ways of working with cache.

It provides `cache` component through GQL context available for all GraphQL Query and Mutation resolver methods.

It also exposes `@cache` directive through the Base GraphQL Schema to the rest of the extensions to define their own caching options.
Such cache will be handled by `@cache` directive itself, so the Extension does not have to do anything by itself in terms of cache checks,

> cache-key generations, etc.

Reviewer: unnecessary enter?

Author: @mateuszwozniak I usually try to keep the line length within a visual part of the IDE (120-140 chars), but such "newlines" have no effects on the page output (markdown does not count it as a paragraph).

## Cache adapters

By default, Falcon-Server uses
[`InMemoryLRUCache`](https://www.apollographql.com/docs/apollo-server/features/data-sources#using-memcachedredis-as-a-cache-storage-backend)

> as a cache backend.

Reviewer: Those line breaks are weird. I assume that's some IDE's auto formatting?

There's an option to configure Redis or Memcached as a cache backend via Falcon-Server config:

<!--DOCUSAURUS_CODE_TABS-->
<!--redis-->

```json
{
  "cache": {
    "type": "redis",
    "options": {
      "host": "redis-server"
    }
  }
}
```

<!--memcached-->

```json
{
  "cache": {
    "type": "memcached",
    "options": ["memcached-server-1", "memcached-server-2"]
  }
}
```

<!--END_DOCUSAURUS_CODE_TABS-->

## Cache Component

> Falcon-Server provides Cache wrapper-component with additional features required for a proper work.

Reviewer 2: "for a proper work"?

Author: changed

As it was mentioned above, this component is accessible via `context` argument in all GraphQL resolvers:

```javascript
async mySampleResolver(parent, args, context, info) {
  // example of getting value for `cache-key` key
  const value = await context.cache.get('cache-key');
}
```

### `get` cache

`async get(key: string, setOptions?: GetCacheOptions): Promise<CacheResult>` method accepts the following arguments:

- `key: string` - cache key
- `setOptions?: GetCacheOptions` - optional object with the following sub-keys:

>   - `fetchData: () => Promise<GetCacheFetchResult>` - this is an optional async fetch-callback, which should return a value if the cached value is missing

Reviewer: ```suggestion
  - `fetchData: () => Promise<GetCacheFetchResult>` - this is an optional async fetch-callback, which should return a value if the cached value is missing. The fetched data will be then cached for a later use.
```

  - `options: SetCacheOptions` - options object to be passed to [`set`](#set-cache) method

> Example of using `keyOrOptions` as an object with `fetchData` callback. Here's a piece of code how you would normally work with the cache:

Reviewer: ```suggestion
Example of using `keyOrOptions` as an object with `fetchData` callback. Here's a piece of code that shows how you would normally work with the cache:
```

- global cache options in app config
- default cache options defined by Base GraphQL Schema

### Caching by tags

When you mark your GraphQL Type with a `@cache` directive, this directive will try to find `ID` field type inside of your
specified type, for example:

```graphql
type Query {
  item(id: ID!): Item @cache
  items: [Item] @cache
}

# Type name is going to be used as a tag prefix
type Item {
  # @cache will check this field to generate a proper tag `Item:id`
  id: ID!
  name: String
}
```

Thus if you run `query { item(id: 1) { id name } }` GQL query, `@cache` will generate `Item` and `Item:1`

> tags automatically for you, and will put the resolver value to the cache with the generated tags for further validation.

Reviewer: ```suggestion
tags automatically for you, and will put the resolved value into the cache with the generated tags for further validation.
```

The same works with a simple list, when you run `query { items { id name } }` - the directive will detect
`[Item]` list and will extract `ID` values from each entry and will generate the following tags -
`Item`, `Item:1`, `Item:2` and so on.

This works great with "plain" values. But when you're dealing with nested listings - you need to help `@cache`

> directive by pointing out where to look for `ID` values at. Let's assume you have the following GQL schema:

Reviewer 2: ```suggestion
directive by pointing out where to look for `ID` values. Let's assume you have the following GQL schema:
```

```graphql
type Query {
  items(page: Int): ItemList @cache(idPath: ["children"])
}

type ItemList {
  children: [Item]
  total: Int
}

type Item {
  id: ID!
  name: String
}
```

> Such `@cache(idPath: ["children"])` statement will tell to the `@cache` directive to look in `children` key

Reviewer: ```suggestion
Such `@cache(idPath: ["children"])` statement will tell to the `@cache` directive to look in `children` property
```
